packages/zoe/src/contracts/loan/updateDebt.js

```diff
@@ -46,32 +50,53 @@ export const makeDebtCalculator = debtCalculatorConfig => {
 
   const config = { ...configMinusGetDebt, getDebt };
 
-  const updateDebt = _state => {
-    const interest = loanMath.make(calcInterestFn(debt.value, interestRate));
-    debt = loanMath.add(debt, interest);
-    debtNotifierUpdater.updateState(debt);
+  const updateDebt = timestamp => {
+    let prevUpdateTimestamp = lastCalculationTimestamp;
+    while (prevUpdateTimestamp + interestPeriod <= timestamp) {
+      prevUpdateTimestamp += interestPeriod;
+      const interest = loanMath.make(calcInterestFn(debt.value, interestRate));
+      debt = loanMath.add(debt, interest);
+      debtNotifierUpdater.updateState(debt);
+    }
     scheduleLiquidation(zcf, config);
+    lastCalculationTimestamp = timestamp;
   };
 
-  /** @type {IterationObserver<undefined>} */
-  const debtObserver = {
-    // Debt is updated with interest every time the
-    // periodAsyncIterable pushes another value
-    updateState: updateDebt,
-    finish: updateDebt,
-    fail: reason => {
+  const addToDebtWhenNotified = lastCount => {
+    const isFinalValue = updateCount => updateCount === undefined;
```

Comment:
people need to understand the notifier abstraction. This helper function just obscures it rather than usefully abstracting it. It would be clearer to just use the `updateCount === undefined` directly